Panic in leader loop when stopping autopilot

### `consul version` for both Client and Server
Client: `v0.8.1`
Server: `v0.8.1`

### `consul info` for both Client and Server
A client is not involved.

Server:
```
agent:
        check_monitors = 0
        check_ttls = 0
        checks = 0
        services = 1
build:
        prerelease =
        revision = 'e9ca44d
        version = 0.8.1
consul:
        bootstrap = false
        known_datacenters = 2
        leader = false
        leader_addr = 10.0.13.24:8300
        server = true
raft:
        applied_index = 482776383
        commit_index = 482776383
        fsm_pending = 0
        last_contact = 48.941638ms
        last_log_index = 482776386
        last_log_term = 2709
        last_snapshot_index = 482775776
        last_snapshot_term = 2709
        latest_configuration = [{Suffrage:Voter ID:10.0.13.24:8300 Address:10.0.13.24:8300} {Suffrage:Voter ID:10.0.13.161:8300 Address:10.0.13.161:8300} {Suffrage:Voter ID:10.0.13.84:8300 Address:10.0.13.84:8300} {Suffrage:Voter ID:10.0.13.231:8300 Address:10.0.13.231:8300} {Suffrage:Voter ID:10.0.13.52:8300 Address:10.0.13.52:8300}]
        latest_configuration_index = 482581927
        num_peers = 4
        protocol_version = 2
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Follower
        term = 2709
runtime:
        arch = amd64
        cpu_count = 8
        goroutines = 204
        max_procs = 2
        os = linux
        version = go1.8.1
serf_lan:
        encrypted = false
        event_queue = 0
        event_time = 400
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 1986
        members = 34
        query_queue = 0
        query_time = 1
serf_wan:
        encrypted = false
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 571
        members = 8
        query_queue = 0
 ```

### Operating system and Environment details

```
NAME="Amazon Linux AMI"
VERSION="2017.03"
ID="amzn"
ID_LIKE="rhel fedora"
VERSION_ID="2017.03"
PRETTY_NAME="Amazon Linux AMI 2017.03"
CPE_NAME="cpe:/o:amazon:linux:2017.03:ga"
```

### Description of the Issue (and unexpected/desired result)
This is same as https://github.com/hashicorp/consul/issues/2896.
Leader node panics when trying to stop autopilot.
The fix in https://github.com/hashicorp/consul/pull/2911 was merged into `0.8.1` release so I'm surprised to see this panic.

### Log Fragments or Link to [gist](https://gist.github.com/)
```
    2017/04/28 13:27:57 [ERR] consul.session: Invalidation failed: leadership lost while committing log
    2017/04/28 13:27:57 [ERR] consul.session: Invalidation failed: leadership lost while committing log
    2017/04/28 13:27:57 [ERR] consul.session: Invalidation failed: leadership lost while committing log
    2017/04/28 13:27:57 [ERR] consul.session: Invalidation failed: leadership lost while committing log
    2017/04/28 13:27:57 [ERR] consul: failed to wait for barrier: leadership lost while committing log
panic: close of closed channel

goroutine 4090476 [running]:
github.com/hashicorp/consul/consul.(*Server).stopAutopilot(0xc420222f00)
        /gopath/src/github.com/hashicorp/consul/consul/autopilot.go:33 +0x2f
github.com/hashicorp/consul/consul.(*Server).revokeLeadership(0xc420222f00, 0xc657305ef8, 0xb0)
        /gopath/src/github.com/hashicorp/consul/consul/leader.go:189 +0x106
github.com/hashicorp/consul/consul.(*Server).leaderLoop(0xc420222f00, 0xc7f92f84e0)
        /gopath/src/github.com/hashicorp/consul/consul/leader.go:121 +0x50f
github.com/hashicorp/consul/consul.(*Server).monitorLeadership.func1(0xc420222f00, 0xc42000c360, 0xc8e89abb30)
        /gopath/src/github.com/hashicorp/consul/consul/leader.go:48 +0x38
created by github.com/hashicorp/consul/consul.(*Server).monitorLeadership
        /gopath/src/github.com/hashicorp/consul/consul/leader.go:50 +0x23b
```

--- a/consul/leader.go
+++ b/consul/leader.go
@@ -65,9 +65,6 @@ func (s *Server) monitorLeadership() {
 // leaderLoop runs as long as we are the leader to run various
 // maintenance activities
 func (s *Server) leaderLoop(stopCh chan struct{}) {
-	// Ensure we revoke leadership on stepdown
-	defer s.revokeLeadership()
-
 	// Fire a user event indicating a new leader
 	payload := []byte(s.config.NodeName)
 	if err := s.serfLAN.UserEvent(newLeaderEvent, payload, false); err != nil {
@@ -101,6 +98,7 @@ RECONCILE:
 			goto WAIT
 		}
 		establishedLeader = true
+		defer s.revokeLeadership()
 	}
 
 	// Reconcile any missing data

--- a/consul/leader.go
+++ b/consul/leader.go
@@ -93,8 +93,7 @@ RECONCILE:
 	// Check if we need to handle initial leadership actions
 	if !establishedLeader {
 		if err := s.establishLeadership(); err != nil {
-			s.logger.Printf("[ERR] consul: failed to establish leadership: %v",
-				err)
+			s.logger.Printf("[ERR] consul: failed to establish leadership: %v", err)
 			goto WAIT
 		}
 		establishedLeader = true
@@ -124,19 +123,20 @@ WAIT:
 			goto RECONCILE
 		case member := <-reconcileCh:
 			s.reconcileMember(member)
-		case <-s.reassertLeaderCh:
-			if establishedLeader {
-				if err := s.revokeLeadership(); err != nil {
-					s.logger.Printf("[ERR] consul: failed to revoke leadership: %v", err)
-					goto WAIT
-				}
-				if err := s.establishLeadership(); err != nil {
-					s.logger.Printf("[ERR] consul: failed to re-establish leadership: %v", err)
-					goto WAIT
-				}
-			}
 		case index := <-s.tombstoneGC.ExpireCh():
 			go s.reapTombstones(index)
+		case <-s.reassertLeaderCh:
+			if !establishedLeader {
+				continue
+			}
+			if err := s.revokeLeadership(); err != nil {
+				s.logger.Printf("[ERR] consul: failed to revoke leadership: %v", err)
+				continue
+			}
+			if err := s.establishLeadership(); err != nil {
+				s.logger.Printf("[ERR] consul: failed to re-establish leadership: %v", err)
+				continue
+			}
 		}
 	}
 }

--- a/consul/leader.go
+++ b/consul/leader.go
@@ -124,6 +124,17 @@ WAIT:
 			goto RECONCILE
 		case member := <-reconcileCh:
 			s.reconcileMember(member)
+		case <-s.reassertLeaderCh:
+			if establishedLeader {
+				if err := s.revokeLeadership(); err != nil {
+					s.logger.Printf("[ERR] consul: failed to revoke leadership: %v", err)
+					goto WAIT
+				}
+				if err := s.establishLeadership(); err != nil {
+					s.logger.Printf("[ERR] consul: failed to re-establish leadership: %v", err)
+					goto WAIT
+				}
+			}
 		case index := <-s.tombstoneGC.ExpireCh():
 			go s.reapTombstones(index)
 		}

--- a/consul/server.go
+++ b/consul/server.go
@@ -174,6 +174,10 @@ type Server struct {
 	// Consul servers.
 	statsFetcher *StatsFetcher
 
+	// reassertLeaderCh is used to signal the leader loop should re-run
+	// leadership actions after a snapshot restore.
+	reassertLeaderCh chan struct{}
+
 	// tombstoneGC is used to track the pending GC invocations
 	// for the KV tombstones
 	tombstoneGC *state.TombstoneGC
@@ -266,6 +270,7 @@ func NewServer(config *Config) (*Server, error) {
 		router:                servers.NewRouter(logger, shutdownCh, config.Datacenter),
 		rpcServer:             rpc.NewServer(),
 		rpcTLS:                incomingTLS,
+		reassertLeaderCh:      make(chan struct{}),
 		tombstoneGC:           gc,
 		shutdownCh:            make(chan struct{}),
 	}

--- a/consul/snapshot_endpoint.go
+++ b/consul/snapshot_endpoint.go
@@ -100,12 +100,10 @@ func (s *Server) dispatchSnapshotRequest(args *structs.SnapshotRequest, in io.Re
 		if err := barrier.Error(); err != nil {
 			return nil, err
 		}
-		if err := s.revokeLeadership(); err != nil {
-			return nil, err
-		}
-		if err := s.establishLeadership(); err != nil {
-			return nil, err
-		}
+
+		// Tell the leader loop to reassert leader actions since we just
+		// replaced the state store contents.
+		s.reassertLeaderCh <- struct{}{}
 
 		// Give the caller back an empty reader since there's nothing to
 		// stream back.

--- a/consul/snapshot_endpoint.go
+++ b/consul/snapshot_endpoint.go
@@ -101,9 +101,14 @@ func (s *Server) dispatchSnapshotRequest(args *structs.SnapshotRequest, in io.Re
 			return nil, err
 		}
 
+		select {
 		// Tell the leader loop to reassert leader actions since we just
 		// replaced the state store contents.
-		s.reassertLeaderCh <- struct{}{}
+		case s.reassertLeaderCh <- struct{}{}:
+
+		// Make sure we don't get stuck during shutdown
+		case <-s.shutdownCh:
+		}
 
 		// Give the caller back an empty reader since there's nothing to
 		// stream back.

